rabbitmq/broker.go

```diff
@@ -0,0 +1,258 @@
+    var headers map[string]interface{}
+    var body []byte
+    var contentType string
+    var contentEncoding string
+    var deliveryMode uint8
+    var correlationId string
+    var replyTo string
+
+    if br.rawMode {
+        headers = make(amqp.Table)
+        body = m
+        contentType = "application/json"
+        contentEncoding = "utf-8"
+        deliveryMode = 2
+        correlationId = ""
+        replyTo = ""
+    } else {
+        var msgmap map[string]interface{}
+        err := json.Unmarshal(m, &msgmap)
+        if err != nil {
+            log.Panicf("Failed to publish a message: %s", err)
+            return err
+        }
+
+        headers = msgmap["headers"].(map[string]interface{})
+        body, err = base64.StdEncoding.DecodeString(msgmap["body"].(string))
+        if err != nil {
+            log.Panicf("Failed to publish a message: %s", err)
+            return err
+        }
+        contentType = msgmap["content-type"].(string)
+        contentEncoding = msgmap["content-encoding"].(string)
+
+        properties_in := msgmap["properties"].(map[string]interface{})
+        deliveryMode = uint8(properties_in["delivery_mode"].(float64))
+        correlationId = properties_in["correlation_id"].(string)
+        replyTo = properties_in["reply_to"].(string)
+    }
+
+    err := br.channel.PublishWithContext(
+        br.ctx,
+        "",     // exchange
+        q,      // routing key
+        false,  // mandatory
+        false,  // immediate
+        amqp.Publishing{
+            Headers: headers,
+            ContentType: contentType,
+            ContentEncoding: contentEncoding,
+            DeliveryMode: deliveryMode,
+            CorrelationId: correlationId,
+            ReplyTo: replyTo,
+            Body: body,
+        })
+    if err != nil {
+        log.Panicf("Failed to publish a message: %s", err)
+    }
+    return err
+}
+
+// Observe sets the queues from which the tasks should be received.
+// Note, the method is not concurrency safe.
+func (br *Broker) Observe(queues []string) {
+    br.queues = queues
+    for _, queue := range queues {
+        _, err := br.channel.QueueDeclare(
+            queue,   // name
+            true,    // durable
+            false,   // delete when unused
+            false,   // exclusive
+            false,   // no-wait
+            nil,     // arguments
+        )
+        if err != nil {
+            log.Panicf("Failed to declare a queue: %s", err)
+        }
+    }
+}
+
+// Receive fetches a Celery task message from a tail of one of the queues in RabbitMQ.
+// After a timeout it returns nil, nil.
+func (br *Broker) Receive() ([]byte, error) {
+
+    const retryIntervalMs = 100
+
+    try_receive := func() (msg amqp.Delivery, ok bool, err error) {
+        queue := br.queues[0]
+        // Put the Celery queue name to the end of the slice for fair processing.
+        broker.Move2back(br.queues, queue)
+        my_msg, my_ok, my_err := br.channel.Get(queue, true)
```

Comment:
Maybe `Consume` is a more convenient way to get messages instead of doing implementing retries here.

Reply:
I looked at that, but I couldn't find a way to get it to return just a single message, while blocking, with a timeout.  Ran out of time, and just implemented what's there for now.  I've worked with RabbitMQ for quite a few years, and IIRC, it's always had this deficiency: no easy way to just fetch the next (single) message from the queue, with blocking, and a timeout.  You can get a batch of messages, but no way to just get one.